File: Sample/BasicSample/sample_cs/Browser/SampleItem.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

class SampleItem : asd.TextureObject2D
{
    public static readonly asd.Vector2DF Size = new asd.Vector2DF(150, 112.5f);
    public static int TitleMaxLength = 9;

    private asd.RectangleShape shape_;
    private asd.TextureObject2D frame { get; set; }

    public ISample Sample { get; private set; }
    public asd.RectangleShape Shape
    {
        get
        {
            if (shape_ == null)
            {
                shape_ = GetArea();
            }
            return shape_;
        }
    }
    public asd.TextObject2D Title { get; private set; }

    public new asd.Vector2DF Position
    {
        get { return base.Position; }
        set
        {
            base.Position = value;
            shape_ = null;
        }
    }

    public SampleItem(ISample sample, asd.Font font)
    {
        Sample = sample;

        Texture = asd.Engine.Graphics.CreateTexture2D("Data/Browser/" + sample.GetType().Name + ".png");
        if (Texture == null)
        {
            Texture = asd.Engine.Graphics.CreateTexture2D("Data/Browser/Default.png");
        }
        Scale = Size / Texture.Size.To2DF();

        string text = "";
        if (sample.Title != "")
        {
            text = sample.Title;
        }
        else
        {
            text = sample.GetType().Name;
        }

        Title = new asd.TextObject2D()
        {
            Text = GetWrappedString(font, text),
            Font = font,
            Position = new asd.Vector2DF(0, 115),
            DrawingPriority = 1,
        };

        frame = new asd.TextureObject2D()
        {
            Texture = asd.Engine.Graphics.CreateTexture2D("Data/Browser/Frame.png"),
            Position = new asd.Vector2DF(-3, -3),
        };
    }

    public asd.RectangleShape GetArea()
    {
        return new asd.RectangleShape
        {
            DrawingArea = new asd.RectF(Position.X, Position.Y, Size.X, Size.Y),
        };
    }

    private string GetWrappedString(asd.Font font, string title)
    {
        string result = title;
        if (font.CalcTextureSize(result, asd.WritingDirection.Horizontal).X <= Size.X)
        {
            return result;
        }
        while (font.CalcTextureSize(result + "…", asd.WritingDirection.Horizontal).X > Size.X)
        {
            result = result.Substring(0, result.Length - 1);
        }
        return result + "…";
    }

    protected override void OnAdded()
    {
        Layer.AddObject(Title);
        Layer.AddObject(frame);

        AddChild(Title, asd.ChildManagementMode.Nothing, asd.ChildTransformingMode.Position);
        AddChild(frame, asd.ChildManagementMode.Nothing, asd.ChildTransformingMode.Position);
    }

    protected override void OnDispose()
    {
        Title.Dispose();
        frame.Dispose();
    }

    public void Activate()
    {
        frame.Texture = asd.Engine.Graphics.CreateTexture2D("Data/Browser/FrameActive.png");
    }

    public void Disactivate()
    {
        frame.Texture = asd.Engine.Graphics.CreateTexture2D("Data/Browser/Frame.png");
    }
}

File: Sample/ApplicationSample/sample_cpp/Browser/SampleItem.cpp
#include "SampleItem.h"

using namespace std;

SampleItem::SampleItem(SampleInfo sample, std::shared_ptr<asd::Font>& font)
	: m_sample(sample)
{
	auto texturePath = asd::ToAString("Data/Browser/") + sample.title + asd::ToAString(".png");
	auto texture = asd::Engine::GetGraphics()->CreateTexture2D(texturePath.c_str());
	if (texture == nullptr)
	{
		texture = asd::Engine::GetGraphics()->CreateTexture2D(asd::ToAString("Data/Browser/Default.png").c_str());
	}
	SetTexture(texture);
	SetScale(asd::Vector2DF(SizeX, SizeY) / texture->GetSize().To2DF());

	asd::astring text;
	if (!sample.readableTitle.empty())
	{
		text = sample.readableTitle;
	}
	else
	{
		text = sample.title;
	}

	m_title = make_shared<asd::TextObject2D>();
	m_title->SetFont(font);
	m_title->SetText(GetWrappedString(font, text).c_str());
	m_title->SetPosition(asd::Vector2DF(0, 115));
	m_title->SetDrawingPriority(1);

	auto frameTexture = asd::Engine::GetGraphics()->CreateTexture2D(asd::ToAString("Data/Browser/Frame.png").c_str());
	m_frame = make_shared<TextureObject2D>();
	m_frame->SetTexture(frameTexture);
	m_frame->SetPosition(asd::Vector2DF(-3, -3));
}

void SampleItem::OnAdded()
{
	AddChild(m_title,
		asd::ChildManagementMode::RegistrationToLayer | asd::ChildManagementMode::Disposal,
		asd::ChildTransformingMode::Position);

	AddChild(m_frame,
		asd::ChildManagementMode::RegistrationToLayer | asd::ChildManagementMode::Disposal,
		asd::ChildTransformingMode::Position);
}

asd::RectangleShape SampleItem::GetArea()
{
	auto result = asd::RectangleShape();
	result.SetDrawingArea(asd::RectF(GetPosition(), asd::Vector2DF(SizeX, SizeY)));
	return result;
}

SampleInfo SampleItem::GetSample() const
{
	return m_sample;
}

void SampleItem::Activate()
{
	auto texture = asd::Engine::GetGraphics()->CreateTexture2D(asd::ToAString("Data/Browser/FrameActive.png").c_str());
	m_frame->SetTexture(texture);
}

void SampleItem::Disactivate()
{
	auto texture = asd::Engine::GetGraphics()->CreateTexture2D(asd::ToAString("Data/Browser/Frame.png").c_str());
	m_frame->SetTexture(texture);
}

asd::astring SampleItem::GetWrappedString(shared_ptr<asd::Font>& font, asd::astring title)
{
	asd::astring result = title;
	if (font->CalcTextureSize(result.c_str(), asd::WritingDirection::Horizontal).X <= SizeX)
	{
		return result;
	}
	while (font->CalcTextureSize(result.c_str(), asd::WritingDirection::Horizontal).X > SizeX)
	{
		result = result.substr(0, result.length() - 1);
	}
	return (result + asd::ToAString("Åc")).c_str();
}